Clear point locator after displacing the mesh

lindsayad opened the pull request:
Displacing the mesh invalidates all the bounding boxes on the
point locator, so we need to make sure we clear the locator so
we don't use bad bounding box data

Closes #22534

lindsayad committed:
Reduce problem size by eliminating nl sys

lindsayad committed:
Demonstrate contains point

lindsayad committed:
Clear the point locator after displacing the mesh

Displacing the mesh invalidates all the bounding boxes on the
point locator, so we need to make sure we clear the locator so
we don't use bad bounding box data

Closes #22534

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1218441/) on 11fae85 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22538/site)

This comment will be updated on new commits.

lindsayad commented:
@maxnezdyur would you be ok with me committing your example as a test with you as the commit author?

roystgnr commented on framework/src/transfers/MultiAppShapeEvaluationTransfer.C:
```diff
@@ -370,6 +370,7 @@ MultiAppShapeEvaluationTransfer::transferVariable(unsigned int i)
         if (node->n_dofs(sys_num, var_num) < 1)
           continue;
 
+        const auto node_id = node->id();
```
What's going on here?  Is this an unused variable?

lindsayad replied:
haha yea that's why CIVET was screaming. It was a temporary debugging line since gdb appears to choke on condition breakpoints involving a call through a range-based for iterator (pointing to a `Node *` in this case)

roystgnr commented on framework/src/loops/UpdateDisplacedMeshThread.C:
```diff
@@ -142,6 +143,12 @@ UpdateDisplacedMeshThread::onNode(NodeRange::const_iterator & nd)
+void
+UpdateDisplacedMeshThread::post()
+{
+  _displaced_problem.mesh().getMesh().clear_point_locator();
```
This seems obviously correct ... which makes it somewhat concerning that it wasn't being done years ago.  Did we never run into a problem until we tried to do a transfer to/from a displaced mesh that relied on a PointLocator query?

lindsayad replied:
It is frankly shocking that no one complained about this before

lindsayad replied:
many kudos to @maxnezdyur for finding it and reporting it

maxnezdyur commented:
> @maxnezdyur would you be ok with me committing your example as a test with you as the commit author?

Yes, sounds good to me.

maxnezdyur committed:
Create test demonstrating incorrect transfer from displaced mesh

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1220020/) on 06e85c2 wanted to post the following:

Warning: This PR changes repo size by 14.67 MiB.

GiudGiud commented:
These files are too big

lindsayad commented:
😢

lindsayad commented:
I tried coarsening the problem to see whether I could reproduce the bad behavior but wasn't successful. I'll try again

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1220053/) on 06e85c2 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>7e9638</th>
      <th colspan=4>#22538 06e85c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.23%</td>
      <td>84.23%</td>
      <td>+0.00%</td>
      <td>93.33%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80283</td>
      <td>80296</td>
      <td>+13</td>
      <td>14</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15028</td>
      <td>15030</td>
      <td>+2</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22538/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22538/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22538/coverage/xfem)

</details>



This comment will be updated on new commits.

maxnezdyur commented:
@lindsayad For a coarser mesh, the deformation needs to be larger for the problem to manifest. The input file below is a modified version of child.i. I changed it to a static mechanics problem and only prescribed displacements that would cause the whole structure to move. You may need need to mess with the timestep size, but I was able to see the problems on a parent mesh with `nx =100 ny=25`. 
```
 [GlobalParams]
  displacements = 'disp_x disp_y'
[]

[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = 0
    ymin = 0
    xmax = 0.2
    ymax = 0.5
    nx = 5
    ny = 15
    elem_type = QUAD4
  []
[]
[Variables]
  [disp_x]
    order = FIRST
  []
  [disp_y]
    order = FIRST
  []
[]

[AuxVariables] # variables that are calculated for output
  [solid_indicator]
    order = FIRST
    family = LAGRANGE
    [AuxKernel]
      type = ConstantAux
      variable = solid_indicator
      value = 0.0
      boundary = 'left right top'
      execute_on = 'INITIAL TIMESTEP_END'
    []
    initial_condition = 1.0
  []
[]

[Modules/TensorMechanics/Master]
  [all]
    # displacements = 'disp_x disp_y'
    add_variables = true
    # new_system = true
    incremental = true
    strain = SMALL
  []
[]

[Materials]
  [elastic_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 10000.0
    poissons_ratio = 0.3
    use_displaced_mesh = true
  []
  [stress]
    type = ComputeFiniteStrainElasticStress
  []
[]

[BCs]
  [move_bottom_x]
    type = FunctionDirichletBC
    boundary = bottom
    function = 't'
    variable = disp_x
  []
  [move_bottom_y]
    type = FunctionDirichletBC
    boundary = bottom
    function = 't'
    variable = disp_y
  []
[]

[Preconditioning]
  [SMP]
    type = SMP
    full = true
  []
[]
[Executioner]
  type = Transient
  end_time = 2
  dt = 1
  solve_type = 'NEWTON'
  petsc_options = '-snes_converged_reason -ksp_converged_reason -snes_ksp_ew'
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_type -pc_factor_shift_type -pc_factor_shift_amount'
  petsc_options_value = 'lu       superlu_dist                  NONZERO               1e-15'
  nl_max_its = 40
  l_max_its = 15
  line_search = 'none'
  nl_abs_tol = 1e-5
  nl_rel_tol = 1e-4
  automatic_scaling = true
[]

[Outputs]
  exodus = true
  print_linear_residuals = false
  print_linear_converged_reason = false
[]

```

lindsayad committed:
Remove untested inputs

lindsayad committed:
Execute multiapps/transfers on initial *and* timestep_begin

In order to get the incorrect results, the point locator must be
first setup in one state with one set of bounding boxes, and then
be used again when the mesh is in a different state

lindsayad committed:
Switch to csvdiff

lindsayad commented:
Thanks @maxnezdyur! Before I saw your comment I switched to a CSVDiff. If that doesn't work out, I will definitely use your coarser child although the biggest file actually comes from the parent output

maxnezdyur commented:
@lindsayad Sounds good. Also, the input file I provided, caused a problem on a parent mesh of nx=100 and ny=25. I didn't supply that input file because that and the time step size were all that I changed in the parent file.

lindsayad commented:
Ah got ya. Ok I think I'm going to switch back to Exodiff with your suggestions as it is a much more reliable illustration of the original problem ... `NodalVariableValue` is a bit hacky as it relies on the fact that the mesh always be numbered the same way

lindsayad committed:
Use coarser meshes and switch back to exodiff

moosebuild commented:
Job [Modules debug](https://civet.inl.gov/job/1220597/) on 5a307f3 : invalidated by @lindsayad
